### PR TITLE
add pod disruption budget for ingress controller

### PR DESCRIPTION
**Pull request by coolbry95**
### Proposed changes
Adds a pod disruption budget resource for the ingress controller pods

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

File: deployments/helm-chart/values.yaml [hunk 1]
```diff
@@ -174,6 +174,16 @@ controller:
   ## The minimum number of seconds for which a newly created Pod should be ready without any of its containers crashing, for it to be considered available.
   minReadySeconds: 0
 
+  ## Pod disruption budget for the Ingress Controller pods.
+  pdb:
```
[the remark this answers is not in the document]
Comment (jasonwilliams14) [reply deployments/helm-chart/values.yaml:178]: Something like:
In template file (more logic would be needed)
```
{{- if .Values.podDisruptionBudget.enabled -}}
```

In values.yaml file (snippet but settings fall under `enabled: false` )

```
podDisruptionBudget:
  enabled: false
```

File: deployments/helm-chart/values.schema.json [hunk 2]
```diff
@@ -1168,6 +1168,41 @@
...
+            },
+            "maxUnavailable": {
+              "type": "integer"
+            }
+          },
+          "examples": [
+            {
+              "enable": true,
+              "minAvailable": 1,
+            }
```
Comment (shaun-nx) [inline deployments/helm-chart/values.schema.json:1199]: JSON is invalid. Missing comma `,` after this curly brace.

**Review by shaun-nx — CHANGES_REQUESTED** [review]

File: deployments/helm-chart/values.schema.json [hunk 3]
```diff
@@ -1168,6 +1168,41 @@
...
+              "type": "integer"
+            },
+            "maxUnavailable": {
+              "type": "integer"
+            }
+          },
+          "examples": [
+            {
+              "enable": true,
+              "minAvailable": 1,
```
Comment (shaun-nx) [inline deployments/helm-chart/values.schema.json:1198]:
```suggestion
              "minAvailable": 1
```

File: deployments/helm-chart/values.schema.json [hunk 4]
```diff
@@ -1168,6 +1168,41 @@
...
+            }
+            {
+              "enable": true,
+              "maxUnavailable": 1,
```
Comment (shaun-nx) [inline deployments/helm-chart/values.schema.json:1202]:
```suggestion
              "maxUnavailable": 1
```

**Review by shaun-nx — CHANGES_REQUESTED** [review]

File: deployments/helm-chart/values.schema.json (hunk 2 above)
Comment (coolbry95) [reply deployments/helm-chart/values.schema.json]: Passes `cat values.schema.json | jq .`

File: deployments/helm-chart/templates/controller-pdb.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+{{- if .Values.controller.podDisruptionBudget.annotations }}
+  annotations:
+{{ toYaml .Values.controller.podDisruptionBudget.annotations | indent 4 }}
+{{- end }}
+spec:
+  selector:
+    matchLabels:
+      app:  {{ include "nginx-ingress.appName" . }}
+{{- if .Values.controller.podDisruptionBudget.minAvailable -}}
+  minAvailable: {{ .Values.controller.pbd.minAvailable }}
```
Comment (vepatel) [inline deployments/helm-chart/templates/controller-pdb.yaml:18]: @coolbry95 
```suggestion
  minAvailable: {{ .Values.controller.podDisruptionBudget.minAvailable }}
```

**Comment by vepatel** [comment]
@coolbry95 setting:
`controller.podDisruptionBudget.maxUnavailable`, `controller.podDisruptionBudget.minAvailable`are giving parse error in `controller-pdb.yaml`  

Output:
`Error: INSTALLATION FAILED: YAML parse error on nginx-ingress/templates/controller-pdb.yaml: error converting YAML to JSON: yaml: line 14: mapping values are not allowed in this context`

File: deployments/helm-chart/templates/controller-pdb.yaml (hunk 5 above)
Comment (coolbry95) [reply deployments/helm-chart/templates/controller-pdb.yaml]: Fixed.

**Comment by coolbry95** [comment]
I have fix the templating in the latest push. Over zealous white space removal.

**Review by vepatel — APPROVED** [review]
👍🏼

File: deployments/helm-chart/values.yaml [hunk 6]
```diff
@@ -179,6 +179,18 @@ controller:
   ## The minimum number of seconds for which a newly created Pod should be ready without any of its containers crashing, for it to be considered available.
   minReadySeconds: 0
 
+  ## Pod disruption budget for the Ingress Controller pods.
+  podDisruptionBudget:
+    ## Enables PodDisruptionBudget.
+    enabled: false
+    ## The annotations of the Ingress Controller pod disruption budget.
+    annotations: {}
+    ## Only one of the two can be specified at once.
```
Comment (jasonwilliams14) [inline deployments/helm-chart/values.yaml:188]:
```suggestion
```

File: deployments/helm-chart/values.yaml [hunk 7]
```diff
@@ -179,6 +179,18 @@ controller:
...
+    ## The number of controller pods that should be available.
```
Comment (jasonwilliams14) [inline deployments/helm-chart/values.yaml:189]:
```suggestion
    ## The number of controller pods that should be available. This is a mutually exclusive setting with “maxUnavailable”.
```

File: deployments/helm-chart/values.yaml [hunk 8]
```diff
@@ -179,6 +179,18 @@ controller:
...
+    #minAvailable: 1
+    ## The number of controller pods that can be unavailable.
```
Comment (jasonwilliams14) [inline deployments/helm-chart/values.yaml:191]:
```suggestion
    ## The number of controller pods that can be unavailable. This is a mutually exclusive setting with “minAvailable”.
```

**Review by jasonwilliams14 — COMMENTED** [review]
Made some suggestions.

File: deployments/helm-chart/values.yaml [hunk 9]
```diff
@@ -179,6 +179,18 @@ controller:
...
 
+  ## Pod disruption budget for the Ingress Controller pods.
+  podDisruptionBudget:
+    ## Enables PodDisruptionBudget.
+    enabled: false
+    ## The annotations of the Ingress Controller pod disruption budget.
+    annotations: {}
+    ## Only one of the two can be specified at once.
+    ## The number of controller pods that should be available.
+    #minAvailable: 1
```
Comment (lucacome) [inline deployments/helm-chart/values.yaml:190]:
```suggestion
    # minAvailable: 1
```

File: deployments/helm-chart/values.yaml [hunk 10]
```diff
@@ -179,6 +179,18 @@ controller:
...
+    ## The number of controller pods that can be unavailable.
+    #maxUnavailable: 1
```
Comment (lucacome) [inline deployments/helm-chart/values.yaml:192]:
```suggestion
    # maxUnavailable: 1
```

File: deployments/helm-chart/values.schema.json [hunk 11]
```diff
@@ -1175,6 +1175,41 @@
...
+              "default": {},
+              "title": "The annotations Schema",
+              "$ref": "https://raw.githubusercontent.com/yannh/kubernetes-json-schema/master/v1.25.4/_definitions.json#/definitions/io.k8s.apimachinery.pkg.apis.meta.v1.ObjectMeta/properties/annotations"
+            },
+            "minAvailable": {
+              "type": "integer"
+            },
+            "maxUnavailable": {
+              "type": "integer"
+            }
```
Comment (lucacome) [inline deployments/helm-chart/values.schema.json:1200]:
```suggestion
            "minAvailable": {
              "$ref": "https://raw.githubusercontent.com/yannh/kubernetes-json-schema/master/v1.25.4/_definitions.json#/definitions/io.k8s.api.policy.v1.PodDisruptionBudgetSpec/properties/minAvailable"
            },
            "maxUnavailable": {
              "$ref": "https://raw.githubusercontent.com/yannh/kubernetes-json-schema/master/v1.25.4/_definitions.json#/definitions/io.k8s.api.policy.v1.PodDisruptionBudgetSpec/properties/maxUnavailable"
            }
```

File: deployments/helm-chart/README.md [hunk 12]
```diff
@@ -256,6 +256,10 @@ Parameter | Description | Default
 `controller.autoscaling.maxReplicas` | Maximum number of replicas for the HPA. | 3
 `controller.autoscaling.targetCPUUtilizationPercentage` | The target CPU utilization percentage. | 50
 `controller.autoscaling.targetMemoryUtilizationPercentage` | The target memory utilization percentage. | 50
+`controller.podDisruptionBudget.enabled` | Enables PodDisruptionBudget. | false
+`controller.podDisruptionBudget.annotations` | The annotations of the Ingress Controller pod disruption budget | {}
+`controller.podDisruptionBudget.minAvailable` | The number of controller pods that should be available. | 0
+`controller.podDisruptionBudget.maxUnavailable` | The number of controller pods that can be unavailable. | 0
```
Comment (lucacome) [inline deployments/helm-chart/README.md:262]:
```suggestion
`controller.podDisruptionBudget.minAvailable` | The number of Ingress Controller pods that should be available. | 0
`controller.podDisruptionBudget.maxUnavailable` | The number of Ingress Controller pods that can be unavailable. | 0
```

File: docs/content/installation/installation-with-helm.md [hunk 13]
```diff
@@ -244,6 +244,10 @@ The following tables lists the configurable parameters of the NGINX Ingress Cont
 |``controller.autoscaling.maxReplicas`` | Maximum number of replicas for the HPA. | 3 |
 |``controller.autoscaling.targetCPUUtilizationPercentage`` | The target CPU utilization percentage. | 50 |
 |``controller.autoscaling.targetMemoryUtilizationPercentage`` | The target memory utilization percentage. | 50 |
+|``controller.podDisruptionBudget.enabled`` | Enables PodDisruptionBudget. | false |
+|``controller.podDisruptionBudget.annotations`` | The annotations of the Ingress Controller pod disruption budget | {} |
+|``controller.podDisruptionBudget.minAvailable`` | The number of controller pods that should be available. | 0 |
+|``controller.podDisruptionBudget.maxUnavailable`` | The number of controller pods that can be unavailable. | 0 |
```
Comment (lucacome) [inline docs/content/installation/installation-with-helm.md:250]:
```suggestion
|``controller.podDisruptionBudget.minAvailable`` | The number of Ingress Controller pods that should be available. | 0 |
|``controller.podDisruptionBudget.maxUnavailable`` | The number of Ingress Controller pods that can be unavailable. | 0 |
```

File: deployments/helm-chart/values.yaml (hunk 8 above)
Comment (lucacome) [reply deployments/helm-chart/values.yaml:191]: `Ingress Controller` here too

File: deployments/helm-chart/values.yaml (hunk 7 above)
Comment (lucacome) [reply deployments/helm-chart/values.yaml:189]: `Ingress Controller` here too

**Review by lucacome — CHANGES_REQUESTED** [review]

File: deployments/helm-chart/README.md [hunk 14]
```diff
@@ -256,6 +256,10 @@ Parameter | Description | Default
 `controller.autoscaling.maxReplicas` | Maximum number of replicas for the HPA. | 3
 `controller.autoscaling.targetCPUUtilizationPercentage` | The target CPU utilization percentage. | 50
 `controller.autoscaling.targetMemoryUtilizationPercentage` | The target memory utilization percentage. | 50
+`controller.podDisruptionBudget.enabled` | Enables PodDisruptionBudget. | false
+`controller.podDisruptionBudget.annotations` | The annotations of the Ingress Controller pod disruption budget | {}
+`controller.podDisruptionBudget.minAvailable` | The number of Ingress Controller pods that should be available. | 0
```
Comment (lucacome) [inline deployments/helm-chart/README.md:261]:
```suggestion
`controller.podDisruptionBudget.minAvailable` | The number of Ingress Controller pods that should be available. This is a mutually exclusive setting with "maxUnavailable". | 0
```

File: deployments/helm-chart/README.md [hunk 15]
```diff
@@ -256,6 +256,10 @@ Parameter | Description | Default
...
+`controller.podDisruptionBudget.maxUnavailable` | The number of Ingress Controller pods that can be unavailable. | 0
```
Comment (lucacome) [inline deployments/helm-chart/README.md:262]:
```suggestion
`controller.podDisruptionBudget.maxUnavailable` | The number of Ingress Controller pods that can be unavailable. This is a mutually exclusive setting with "minAvailable". | 0
```

**Review by lucacome — APPROVED** [review]
Sorry just seeing this.

**Comment by coolbry95** [comment]
Fixed the merge conflict.

File: deployments/helm-chart/README.md [hunk 16]
```diff
@@ -256,7 +256,11 @@ Parameter | Description | Default
...
 `controller.autoscaling.targetMemoryUtilizationPercentage` | The target memory utilization percentage. | 50
-`controller.strategy` | Specifies the strategy used to replace old Pods with new ones. Docs for [Deployment update strategy](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#strategy) and [Daemonset update strategy](https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#daemonset-update-strategy) | {}
+`controller.podDisruptionBudget.enabled` | Enables PodDisruptionBudget. | false
+`controller.podDisruptionBudget.annotations` | The annotations of the Ingress Controller pod disruption budget | {}
+`controller.podDisruptionBudget.minAvailable` | The number of Ingress Controller pods that should be available. This is a mutually exclusive setting with "maxUnavailable". | 0
+`controller.podDisruptionBudget.maxUnavailable` | The number of Ingress Controller pods that can be unavailable. This is a mutually exclusive setting with "minAvailable". | 0
+`controller.strategy` | Specifies the strategy used to replace old Pods by new ones. [docs](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#strategy) | {}
```
Comment (lucacome) [inline deployments/helm-chart/README.md:263]: I think something happened here from a rebase 🤔

Comment (coolbry95) [reply deployments/helm-chart/README.md]: Fixed. I don't use merges locally. It is throwing me off on this project.

**Comment by vepatel** [comment]
merge away @coolbry95 once this run passes!

**Comment by coolbry95** [comment]
> merge away @coolbry95 once this run passes!

I do not have merge privileges.

**Review by jasonwilliams14 — APPROVED** [review]

**Pull request merged**